collect_app/src/main/java/org/odk/collect/android/application/initialization/SavepointsImporter.kt

```diff
@@ -0,0 +1,75 @@
+import org.odk.collect.android.utilities.SavepointsRepositoryProvider
+import org.odk.collect.forms.FormsRepository
+import org.odk.collect.forms.instances.InstancesRepository
+import org.odk.collect.forms.savepoints.Savepoint
+import org.odk.collect.forms.savepoints.SavepointsRepository
+import org.odk.collect.projects.ProjectsRepository
+import org.odk.collect.settings.keys.MetaKeys
+import org.odk.collect.upgrade.Upgrade
+import java.io.File
+
+class SavepointsImporter(
+    private val projectsRepository: ProjectsRepository,
+    private val instancesRepositoryProvider: InstancesRepositoryProvider,
+    private val formsRepositoryProvider: FormsRepositoryProvider,
+    private val savepointsRepositoryProvider: SavepointsRepositoryProvider,
+    private val storagePathProvider: StoragePathProvider
+) : Upgrade {
+    override fun key(): String {
+        return MetaKeys.OLD_SAVEPOINTS_IMPORTED
+    }
+
+    override fun run() {
+        projectsRepository.getAll().forEach { project ->
+            val instancesRepository = instancesRepositoryProvider.get(project.uuid)
+            val formsRepository = formsRepositoryProvider.get(project.uuid)
+            val savepointsRepository = savepointsRepositoryProvider.get(project.uuid)
+            val cacheDir = File(storagePathProvider.getOdkDirPath(StorageSubdirectory.CACHE, project.uuid))
+            val instancesDir = File(storagePathProvider.getOdkDirPath(StorageSubdirectory.INSTANCES, project.uuid))
+
+            importSavepointsThatBelongToSavedForms(instancesRepository, formsRepository, savepointsRepository, cacheDir)
+
+            importSavepointsThatBelongToBlankForms(formsRepository, savepointsRepository, cacheDir, instancesDir)
+        }
+    }
+
+    private fun importSavepointsThatBelongToSavedForms(instancesRepository: InstancesRepository, formsRepository: FormsRepository, savepointsRepository: SavepointsRepository, cacheDir: File) {
+        instancesRepository.all.forEach { instance ->
+            if (instance.deletedDate == null) {
+                val savepointFile = File(cacheDir, File(instance.instanceFilePath).name.plus(".save"))
+                if (savepointFile.exists() && savepointFile.lastModified() > instance.lastStatusChangeDate) {
+                    formsRepository.getAllByFormIdAndVersion(instance.formId, instance.formVersion).firstOrNull()?.let { form ->
+                        if (!form.isDeleted) {
+                            savepointsRepository.save(Savepoint(form.dbId, instance.dbId, savepointFile.absolutePath, instance.instanceFilePath))
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun importSavepointsThatBelongToBlankForms(formsRepository: FormsRepository, savepointsRepository: SavepointsRepository, cacheDir: File, instancesDir: File) {
+        formsRepository.all.sortedByDescending { form -> form.date }.forEach { form ->
```

Comment:
I sort the forms here so that we start looking for savepoints for the newest ones. Thanks to that newer versions of the same forms will be taken into account first which is important because newer forms usually have the same name like the first one but we add `_1`, `_2`, etc. suffixes:
`form.xml (v1)`
`form_2.xml (v2)`
Savepoints for such forms would be named:
`form_2024-04-10_01-35-41.xml.save`
`form_1_2024-04-10_01-45-41.xml.save`
If we start looking for savepoints for older form versions first we could end up with a wrong savepoint loaded. For example, if we started with `form.xml (v1)` we would look for a file with a name starting with `form_` then both savepoints `form_2024-04-10_01-35-41.xml.save` and `form_1_2024-04-10_01-45-41.xml.save` would match and the one that belongs to the newer version of this form could be used.